### ISSUE TITLE
testing: add test for git-clonerepos.yml playbook

### ISSUE THREAD

**Issue by sometheycallme** [cleanerbot/ansible-security#81]
@Jfach @jumanjiman @jwasserzug @vpalacio 

#73 magic variable {{ inventory_dir }} works for vars_files in playbooks

#80 os.putenv() works for creating ANSIBLE_HOSTS (the inventory_dir)

In server.py we use os.putenv() to create the staged directory with data passed from the contributor to the ansible-controller 

```python
            command += " {0}/{1}/{2}".format(base_dir, safe_dir, playbook)
            print command
            # Use this to set ANSIBLE_HOSTS environment variable as base_dir, safe_dir
            # root is hosts
            os.putenv('ANSIBLE_HOSTS', "{0}/{1}/hosts".format(base_dir, safe_dir)
            # change directory to contributors staged directory to run commands
            os.chdir("/{0}/{1}".format(base_dir, safe_dir))
            os.system(command)
```

**Comment by sometheycallme** [comment]
```shell

user@devenv:~/ansible-security(master)$ ansible --version
ansible 2.1.0.0
  config file = /home/user/ansible-security/ansible.cfg

```

**Comment by jumanjiman** [comment]
well-done!

**Comment by sometheycallme** [comment]
@jumanjiman - thanks - getting closer.  Could use your help though.

```
  File "webserver/server.py", line 79
    os.chdir("/{0}/{1}".format(base_dir, safe_dir))
     ^
SyntaxError: invalid syntax
```

Gotta fix this next for testing.

**Comment by jumanjiman** [comment]
perhaps it would be clearer and more portable to do something like:

```python
path = os.path.join(base_dir, safe_dir)
os.chdir(path)
```

**Comment by sometheycallme** [comment]
Yes, I agree.

Trying this:


https://github.com/cleanerbot/ansible-security/pull/82/commits/31e6edfec42933d21b9af20546de60adcb91b752

**Comment by sometheycallme** [comment]
Python seems to like it, will try via curl next.

```
user@devenv:~/ansible-security(test-gitrepo)$ python webserver/server.py 
Listening on localhost:8080
```

**Comment by sometheycallme** [comment]
Merged and rebuilding pipeline.  Will spawn another PR.

**Comment by sometheycallme** [comment]
Now to be sure I'm keeping track (steps to replicate test of pipeline components once up and running)

```
core@pipeline ~ $ docker ps
CONTAINER ID        IMAGE                     COMMAND                  CREATED             STATUS              PORTS                     NAMES
7947bbdd3abc        autostager                "python autostager/au"   2 minutes ago       Up 2 minutes                                  adoring_dijkstra
0d6b34f8fc55        ansible-controller        "python webserver/ser"   2 minutes ago       Up 2 minutes        0.0.0.0:8080->8080/tcp    small_jang
```

docker inspect 0d6b34f8fc55 | grep IPAddress

yields:
```
"IPAddress": "172.17.0.5",
```

autostager has reported to have staged the request:

```
Proccessing cleanerbot:test-gitrepo
=========================
    INIT PULL REQUEST    
=========================
Branch: test-gitrepo
Clone URL: https://111a4a337a9e88bae7d98bfd4a781f4daabec9b2@github.com/cleanerbot/ansible-security.git
Base Dir: /opt/staging
Name: cleanerbot_test_gitrepo
```

Without using helpful_files/play.py (yet) pass curl to the webserver:

```json
curl -v -X POST -d '{"branch_name": "test-gitrepo", "git_handle": "cleanerbot", "flags": [{"flag": "-i", "argument": "hosts"}], "playbook": "playbooks/git-clonerepos.yml"}' http://172.17.0.5:8080/play
```

yields this from the webserver on controller

```
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 172.17.0.5...
* Connected to 172.17.0.5 (172.17.0.5) port 8080 (#0)
> POST /play HTTP/1.1
> Host: 172.17.0.5:8080
> User-Agent: curl/7.47.1
> Accept: */*
> Content-Length: 151
> Content-Type: application/x-www-form-urlencoded
> 
* upload completely sent off: 151 out of 151 bytes
* Empty reply from server
* Connection #0 to host 172.17.0.5 left intact
curl: (52) Empty reply from server
```



Playbook doesn't run.  Next thing to fix.....

**Comment by sometheycallme** [comment]
Interesting:

docker exec -i -t 0d6b34f8fc55 bash

```
bash-4.3# cd cleanerbot_test_gitrepo/
bash-4.3# ls
LICENSE        README.md      assets         dockerfiles    group_vars     hosts          script
Makefile       ansible.cfg    circle.yml     fixtures       helpful_files  playbooks      webserver

bash-4.3# 
bash-4.3# ansible-playbook -i hosts playbooks/git-clonerepos.yml 

PLAY [Clone git repos] *********************************************************

TASK [Cloning repos] ***********************************************************
failed: [localhost] (item={u'repo': u'docker-nessus', u'org': u'cleanerbot'}) => {"failed": true, "item": {"org": "cleanerbot", "repo": "docker-nessus"}, "msg": "Failed to find required executable ssh-keyscan"}

NO MORE HOSTS LEFT *************************************************************
 [ERROR]: Could not create retry file 'playbooks/git-clonerepos.retry'. The error was: [Errno 30] Read-only file system: 'playbooks/git-
clonerepos.retry'

PLAY RECAP *********************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1   

bash-4.3# 
```

couple of things we'll need to fix there, including the read-only filesystem.  looks like the controller will need a read-write place for the .retry files.

**Comment by jumanjiman** [comment]
can you override the location of the retry files?
if so, you could then use `docker run --tmpfs=/some/path ...`

**Comment by sometheycallme** [comment]
>can you override the location of the retry files?

looks like there are a few options wrt retry files, including turning them off altogether.

http://stackoverflow.com/questions/31318881/how-do-you-stop-ansible-from-creating-retry-files-in-the-home-directory

**Comment by greenpau** [comment]
@sometheycallme , do you have a PR open for this one? I would like to comment on the code.

**Comment by sometheycallme** [comment]
>do you have a PR open for this one

@greenpau - hi.

See this for WIP.

https://github.com/cleanerbot/ansible-security/pull/83

**Comment by sometheycallme** [comment]
FYI - we are using the ```/ansible-security/playbooks/git-clonerepos.yml`` as the candidate playbook for CI.

What I was attempting to do here is spawn the entire devenv pipeline (in Digital Ocean) and pass the curl command over to the ansible-controller - only after the branch was staged.

**Comment by sometheycallme** [comment]
You can see that ```autostager``` - it stages the branch & PR for the controller ( comment by "kropped").   The ansible-controller subsequently has the /opt/staging/<PR info>

@greenpau please let me know if you would like a devenv in digital ocean as well.  I can build one for you rather quickly and add you to DUO

**Comment by sometheycallme** [comment]
self.reminder()

Writing a test for this playbook validates two things with the pipeline

1) ANSIBLE_HOSTS is set properly

2) Nested variables work for a playbook when ANSIBLE_HOSTS is setup properly.

Failed test means that /opt/staging directory and/or nested variables are not working properly.
